prow/pubsub/subscriber/subscriber.go

```diff
@@ -148,265 +130,48 @@ func (m *pubSubMessage) nack() {
+				l.WithError(err).Warning("Failed to report status.")
 			}
-			postsubmitJob = &job
 		}
 	}
-	// This also captures the case where fetching jobs from inrepoconfig failed.
-	// However doesn't not distinguish between this case and a wrong prow job name.
-	if postsubmitJob == nil {
-		err = fmt.Errorf("failed to find associated postsubmit job %q from orgRepo %q", pe.Name, orgRepo)
-		return
-	}
-
-	spec := pjutil.PostsubmitSpec(*postsubmitJob, *refs)
-	jobSpec, labels, annotations = &spec, postsubmitJob.Labels, postsubmitJob.Annotations
-	return
-}
-
-func extractFromAttribute(attrs map[string]string, key string) (string, error) {
-	value, ok := attrs[key]
-	if !ok {
-		return "", fmt.Errorf("unable to find %q from the attributes", key)
-	}
-	return value, nil
 }
 
 func (s *Subscriber) handleMessage(msg messageInterface, subscription string, allowedClusters []string) error {
-	// Observed message payload drifts to the payload from another pubsub
-	// message and couldn't figure out why. Extracts the values ahead of time
-	// and hopefully could mitigate this issue for now.
-	// TODO(chaodaiG): remove these once figured out the root cause of data race.
-	msgID := msg.getID()
-	msgPayload := msg.getPayload()
-	msgAttributes := msg.getAttributes()
 
+	msgID := msg.getID()
 	l := logrus.WithFields(logrus.Fields{
 		"pubsub-subscription": subscription,
 		"pubsub-id":           msgID})
-	s.Metrics.MessageCounter.With(prometheus.Labels{subscriptionLabel: subscription}).Inc()
 
-	l.WithField("payload", string(msgPayload)).Debug("Received message")
-	eType, err := extractFromAttribute(msgAttributes, ProwEventType)
+	// First, convert the incoming message into a CreateJobExecutionRequest type.
+	cjer, err := s.msgToCjer(l, msg, subscription)
 	if err != nil {
-		l.WithError(err).Error("failed to read message")
-		s.Metrics.ErrorCounter.With(prometheus.Labels{
-			subscriptionLabel: subscription,
-			errorTypeLabel:    "malformed-message",
-		}).Inc()
 		return err
 	}
 
-	var jh jobHandler
-	switch eType {
-	case PeriodicProwJobEvent:
-		jh = &periodicJobHandler{}
-	case PresubmitProwJobEvent:
-		jh = &presubmitJobHandler{}
-	case PostsubmitProwJobEvent:
-		jh = &postsubmitJobHandler{}
-	default:
-		l.WithField("type", eType).Info("Unsupported event type")
-		s.Metrics.ErrorCounter.With(prometheus.Labels{
-			subscriptionLabel: subscription,
-			errorTypeLabel:    "unsupported-event-type",
-		}).Inc()
-		return fmt.Errorf("unsupported event type: %s", eType)
-	}
-	if err = s.handleProwJob(l, jh, msgPayload, subscription, eType, allowedClusters); err != nil {
+	// Do not check for HTTP client authorization, because we're handling a
+	// PubSub message.
+	var allowedApiClient *config.AllowedApiClient = nil
+	var requireTenantID bool = false
+
+	if _, err = gangway.HandleProwJob(l, s.getReporterFunc(l), cjer, s.ProwJobClient, s.ConfigAgent.Config(), s.InRepoConfigCacheHandler, allowedApiClient, requireTenantID, allowedClusters); err != nil {
```

Comment:
glad to see that the pubsub logics are replaced by Gangway, I have two concerns here:

- Users are expected to get pubsub notifications when they trigger via pubsub, I think there are a few places in the file that pubsub message are expected to be sent back but seem to be omitted in this refactoring
- This PR is already pretty large, could we leave the change in pubsub to a separate PR so that the potential risk is lower?

Reply:
>  I think there are a few places in the file that pubsub message are expected to be sent back but seem to be omitted in this refactoring

Can you be more specific?

Reply:
The new `gangway.HandleProwJob` method takes a `reporterFunc`, and Sub passes in its own `getReporterFunc` so that the reporting-back behavior is preserved AFAICS. Was there some other thing you had in mind?

Reply:
ack. like we had discussed offline, that I would suggest breaking this up into 2 separate PRs, as there are users actively use the pubsub functions. So if in case there is any bug on the pubsub or gangway side we could just revert the 2nd PR. However, if you could make the commit history of this PR clean enough that could also work